File: 2025-2/sesion-de-clase-09/SoftProg/SoftProgPersistencia/src/main/java/pe/edu/pucp/inf30/softprog/daoimpl/clientes/ClienteDAOImpl.java
package pe.edu.pucp.inf30.softprog.daoimpl.clientes;

import java.sql.CallableStatement;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Types;
import pe.edu.pucp.inf30.softprog.dao.clientes.ClienteDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.BaseDAO;
import pe.edu.pucp.inf30.softprog.daoimpl.rrhh.CuentaUsuarioDAOImpl;
import pe.edu.pucp.inf30.softprog.modelo.Genero;
import pe.edu.pucp.inf30.softprog.modelo.clientes.CategoriaCliente;
import pe.edu.pucp.inf30.softprog.modelo.clientes.Cliente;

/**
 *
 * @author eric
 */
public class ClienteDAOImpl extends BaseDAO<Cliente> implements ClienteDAO {
    @Override
    protected PreparedStatement comandoCrear(Connection conn, Cliente modelo) 
            throws SQLException {
        
        String sql = "{call insertarCliente(?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_idCuentaUsuario", modelo.getCuentaUsuario().getId());
        cmd.setString("p_dni", modelo.getDni());
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_apellidoPaterno", modelo.getApellidoPaterno());
        cmd.setString("p_genero", String.valueOf(modelo.getGenero()));
        cmd.setDate("p_fechaNacimiento", new java.sql.Date(
                modelo.getFechaNacimiento().getTime()));
        cmd.setString("p_categoria", modelo.getCategoria().name());
        cmd.setDouble("p_lineaCredito", modelo.getLineaCredito());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.registerOutParameter("p_id", Types.INTEGER);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoActualizar(Connection conn, 
            Cliente modelo) throws SQLException {
        
        String sql = "{call modificarCliente(?, ?, ?, ?, ?, ?, ?, ?, ?, ?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_idCuentaUsuario", modelo.getCuentaUsuario().getId());
        cmd.setString("p_dni", modelo.getDni());
        cmd.setString("p_nombre", modelo.getNombre());
        cmd.setString("p_apellidoPaterno", modelo.getApellidoPaterno());
        cmd.setString("p_genero", String.valueOf(modelo.getGenero()));
        cmd.setDate("p_fechaNacimiento", 
                new java.sql.Date(modelo.getFechaNacimiento().getTime()));
        cmd.setString("p_categoria", modelo.getCategoria().name());
        cmd.setDouble("p_lineaCredito", modelo.getLineaCredito());
        cmd.setBoolean("p_activo", modelo.isActivo());
        cmd.setInt("p_id", modelo.getId());
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoEliminar(Connection conn, Integer id) 
            throws SQLException {
        
        String sql = "{call eliminarCliente(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeer(Connection conn, Integer id) 
            throws SQLException {
        
        String sql = "{call buscarClientePorId(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setInt("p_id", id);
        
        return cmd;
    }

    @Override
    protected PreparedStatement comandoLeerTodos(Connection conn) 
            throws SQLException {
        
        String sql = "{call listarClientes()}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        
        return cmd;
    }

    @Override
    protected Cliente mapearModelo(ResultSet rs) throws SQLException {
        Cliente cliente = new Cliente();
        cliente.setId(rs.getInt("id"));
        cliente.setCuentaUsuario(
                new CuentaUsuarioDAOImpl().leer(rs.getInt("idCuentaUsuario")));
        cliente.setDni(rs.getString("dni"));
        cliente.setNombre(rs.getString("nombre"));
        cliente.setApellidoPaterno(rs.getString("apellidoPaterno"));
        cliente.setGenero(Genero.valueOf(rs.getString("genero")));
        cliente.setFechaNacimiento(rs.getTimestamp("fechaNacimiento"));
        cliente.setLineaCredito(rs.getDouble("lineaCredito"));
        cliente.setCategoria(
                CategoriaCliente.valueOf(rs.getString("categoria")));
        cliente.setActivo(rs.getBoolean("activo"));
        
        return cliente;
    }

    protected PreparedStatement comandoBuscarPorDni(Connection conn, String dni) 
            throws SQLException {
        
        String sql = "{call buscarClientePorDni(?)}";
        
        CallableStatement cmd = conn.prepareCall(sql);
        cmd.setString("p_dni", dni);
        
        return cmd;
    }
    
    @Override
    public Cliente buscarPorDni(String dni) {
        return ejecutarComando(conn -> {
            try (PreparedStatement cmd = this.comandoBuscarPorDni(conn, dni)) {
                ResultSet rs = cmd.executeQuery();

                if (!rs.next()) {
                    System.err.println("No se encontro el registro con "
                            + "dni: " + dni);
                    return null;
                }

                return this.mapearModelo(rs);
            }
        });
    }
}

File: 2025-2/sesion-de-clase-11/SoftProgPersistencia/DAOImpl/Clientes/ClienteDAOImpl.cs
using System;
using System.Data;
using System.Data.Common;
using PUCP.SoftProg.Modelo.Clientes;
using PUCP.SoftProg.Persistencia.DAO.Clientes;
using PUCP.SoftProg.Modelo;

namespace PUCP.SoftProg.Persistencia.DAOImpl.Clientes {
    public class ClienteDAOImpl : BaseDAOImpl<Cliente>, IClienteDAO {
        protected override DbCommand ComandoCrear(DbConnection conn, Cliente cliente) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "insertarCliente";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32, cliente.CuentaUsuario.Id);
            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.String, cliente.Dni);
            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, cliente.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_apellidoPaterno", DbType.String, cliente.ApellidoPaterno);
            this.AgregarParametroEntrada(cmd, "@p_genero", DbType.Double, cliente.Genero);
            this.AgregarParametroEntrada(cmd, "@p_fechaNacimiento", DbType.DateTime, cliente.FechaNacimiento);
            this.AgregarParametroEntrada(cmd, "@p_categoria", DbType.String, cliente.Categoria);
            this.AgregarParametroEntrada(cmd, "@p_lineaCredito", DbType.Double, DBNull.Value);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, cliente.IsActive);
            this.AgregarParametroSalida(cmd, "@p_id", DbType.Int32);

            return cmd;
        }

        protected override DbCommand ComandoActualizar(DbConnection conn, Cliente cliente) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "modificarCliente";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_idCuentaUsuario", DbType.Int32, cliente.CuentaUsuario.Id);
            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.String, cliente.Dni);
            this.AgregarParametroEntrada(cmd, "@p_nombre", DbType.String, cliente.Nombre);
            this.AgregarParametroEntrada(cmd, "@p_apellidoPaterno", DbType.String, cliente.ApellidoPaterno);
            this.AgregarParametroEntrada(cmd, "@p_genero", DbType.Double, cliente.Genero);
            this.AgregarParametroEntrada(cmd, "@p_fechaNacimiento", DbType.DateTime, cliente.FechaNacimiento);
            this.AgregarParametroEntrada(cmd, "@p_categoria", DbType.String, cliente.Categoria);
            this.AgregarParametroEntrada(cmd, "@p_lineaCredito", DbType.Double, DBNull.Value);
            this.AgregarParametroEntrada(cmd, "@p_activo", DbType.Boolean, cliente.IsActive);
            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, cliente.Id);

            return cmd;
        }

        protected override DbCommand ComandoEliminar(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "eliminarCliente";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeer(DbConnection conn, int id) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarClientePorId";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_id", DbType.Int32, id);

            return cmd;
        }

        protected override DbCommand ComandoLeerTodos(DbConnection conn) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "listarClientes";
            cmd.CommandType = CommandType.StoredProcedure;
            return cmd;
        }

        protected override Cliente MapearModelo(DbDataReader reader) {
            return new Cliente {
                Id = Convert.ToInt32(reader["id"]),
                Dni = Convert.ToString(reader["dni"]),
                Nombre = Convert.ToString(reader["nombre"]),
                ApellidoPaterno = Convert.ToString(reader["apellidoPaterno"]),
                Genero = (Genero)Enum.Parse(typeof(Genero), Convert.ToString(reader["genero"])),
                FechaNacimiento = Convert.ToDateTime(reader["fechaNacimiento"]),
                Categoria = (CategoriaCliente)Enum.Parse(typeof(CategoriaCliente), 
                    Convert.ToString(reader["categoria"])),
                IsActive = Convert.ToBoolean(reader["activo"])
            };
        }

        protected DbCommand ComandoBuscarPorDni(DbConnection conn, string dni) {
            DbCommand cmd = conn.CreateCommand();
            cmd.CommandText = "buscarClientePorDni";
            cmd.CommandType = CommandType.StoredProcedure;

            this.AgregarParametroEntrada(cmd, "@p_dni", DbType.Int32, dni);

            return cmd;
        }

        public Cliente BuscarPorDni(string dni) {
            return EjecutarComando(conn => {
                using (DbCommand cmd = this.ComandoBuscarPorDni(conn, dni))
                using (DbDataReader reader = cmd.ExecuteReader()) {
                    if (!reader.HasRows) {
                        Console.Error.WriteLine("No se encontro el cliente con dni: " + dni);
                        return default;
                    }

                    reader.Read();
                    return this.MapearModelo(reader);
                }
            });
        }
    }
}
